transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamRequestManager.java

```diff
@@ -39,104 +46,204 @@ final class ClientStreamRequestManager<M extends BufferWriter> {
+    final var payload = BufferUtil.bufferAsArray(request);
+
+    servers.forEach(
+        serverId ->
+            communicationService.unicast(
+                StreamTopics.REMOVE.topic(), payload, Function.identity(), serverId, true));
+  }
+
+  @VisibleForTesting("Allows easier test set up and validation")
+  ClientStreamRegistration<M> registrationFor(
+      final AggregatedClientStream<M> stream, final MemberId serverId) {
+    final var streamsPerHost = registrations.computeIfAbsent(serverId, ignored -> new HashMap<>());
+    return streamsPerHost.computeIfAbsent(
+        stream.getStreamId(), streamId -> new ClientStreamRegistration<>(stream, serverId));
+  }
+
+  private void sendAddRequest(
+      final ClientStreamRegistration<M> registration, final byte[] request) {
+    if (registration.state() != State.ADDING) {
+      return;
+    }
+
+    final var pendingRequest =
         communicationService.send(
             StreamTopics.ADD.topic(),
             request,
-            BufferUtil::bufferAsArray,
             Function.identity(),
-            brokerId,
+            Function.identity(),
+            registration.serverId(),
             REQUEST_TIMEOUT);
-
-    result.whenCompleteAsync(
-        (ignored, error) -> {
-          if (error != null) {
-            LOG.warn(
-                "Failed to open stream {} to node {}. Will retry in {}",
-                clientStream,
-                brokerId,
-                RETRY_DELAY,
-                error);
-            // TODO: define some abort conditions. We may not have to retry indefinitely.
-            // For now we retry always. Eventually the request will succeed. Duplicate add request
-            // are fine.
-            executor.schedule(RETRY_DELAY, () -> doAdd(request, brokerId, clientStream));
-          } else {
-            LOG.debug("Opened stream {} to node {}", clientStream, brokerId);
-            clientStream.add(brokerId);
-          }
-        },
-        executor::run);
+    registration.setPendingRequest(pendingRequest);
+    pendingRequest.whenCompleteAsync(
+        (ok, error) -> handleAddResponse(registration, request, error), executor::run);
   }
 
-  void removeStream(
-      final AggregatedClientStream<M> clientStream, final Collection<MemberId> servers) {
-    final var request = new RemoveStreamRequest().streamId(clientStream.getStreamId());
-    servers.forEach(brokerId -> executor.run(() -> doRemove(request, brokerId, clientStream)));
+  private void handleAddResponse(
+      final ClientStreamRegistration<M> registration, final byte[] request, final Throwable error) {
+    final var state = registration.state();
+    if (state != State.ADDING) {
+      LOGGER.trace("Skip handling ADD response since the state is {}", state, error);
+      return;
+    }
+
+    if (error == null) {
+      registration.transitionToAdded();
+      return;
+    }
+
+    // TODO: define some abort conditions. We may not have to retry indefinitely.
```

Comment:
:wrench: I think this comment can be removed. Now the request will be retried only until the registration is marked as removing/removed.